man/dmexport.1
.TH "DMEXPORT" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmexport \(en exports records from database to file
.SH SYNOPSIS
\f[B]dmexport\f[R] \-\-help
.PP
\f[B]dmexport\f[R] \-\-version
.PP
\f[B]dmexport\f[R] \-\-\f[B]database\f[R] \f[I]file\f[R]
\-\-\f[B]type\f[R] \f[I]type\f[R] \-\-\f[B]format\f[R] \f[I]format\f[R]
\-\-\f[B]node\f[R] \f[I]id\f[R] [\-\-\f[B]output\f[R] \f[I]file\f[R]]
[\-\-\f[B]sensor\f[R] \f[I]id\f[R]] [\-\-\f[B]target\f[R] \f[I]id\f[R]]
[\-\-\f[B]response\f[R] \f[I]name\f[R]] [\-\-\f[B]from\f[R]
\f[I]timestamp\f[R]] [\-\-\f[B]to\f[R] \f[I]timestamp\f[R]]
[\-\-\f[B]header\f[R]]
.SH DESCRIPTION
The \f[B]dmexport\f[R] program writes logs, nodes, sensors, targets, and
observations from database to file, in ASCII block, CSV, JSON, JSON
Lines, Fortran 95 Namelist format.
The ASCII block data format is only available for X/Y data points.
.PP
The types data point, log, and observation require a given sensor id,
target id, and time range in ISO 8601 format.
.PP
If no output file is given, the data is printed to standard output.
The output file will be overwritten if it already exists.
If no records are found, an empty file will be created.
.SH OPTIONS
.TP
\f[B]\-\-database\f[R], \f[B]\-d\f[R] \f[I]file\f[R]
Path of the SQLite database (required).
.TP
\f[B]\-\-format\f[R], \f[B]\-f\f[R] [block|csv|json|jsonl|nml|tsv]
Output file format, either ASCII block, CSV, JSON, JSON Lines, NML, or
TSV (required).
Only data points can be exported in block format.
.TP
\f[B]\-\-from\f[R], \f[B]\-B\f[R] \f[I]timestamp\f[R]
Start of time range in ISO 8601 (required for types \f[CR]db\f[R],
\f[CR]log\f[R], \f[CR]observ\f[R]).
.TP
\f[B]\-\-header\f[R], \f[B]\-H\f[R]
Enable CSV/TSV header.
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Node id (required).
.TP
\f[B]\-\-output\f[R], \f[B]\-o\f[R] \f[I]file\f[R]
Path of output file.
Empty or \f[CR]\-\f[R] for standard output.
.TP
\f[B]\-\-response\f[R], \f[B]\-R\f[R] \f[I]name\f[R]
Response name (required for type \f[CR]dp\f[R]).
.TP
\f[B]\-\-sensor\f[R], \f[B]\-S\f[R] \f[I]id\f[R]
Sensor id (required for types \f[CR]dp\f[R] and \f[CR]observ\f[R]).
.TP
\f[B]\-\-separator\f[R], \f[B]\-s\f[R] \f[I]char\f[R]
Single character used as separator instead of comma.
.TP
\f[B]\-\-target\f[R], \f[B]\-T\f[R] \f[I]id\f[R]
Target id (required for type \f[CR]dp\f[R] and \f[CR]observ\f[R]).
.TP
\f[B]\-\-to\f[R], \f[B]\-E\f[R] \f[I]timestamp\f[R]
End of time range in ISO 8601 (required for types \f[CR]dp\f[R],
\f[CR]log\f[R], and \f[CR]observ\f[R]).
.TP
\f[B]\-\-type\f[R], \f[B]\-t\f[R] [beat|dp|log|node|observ|sensor|target]
Type of records to export (required).
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Records have been exported.
.TP
\f[B]1\f[R]
Failure.
Export of records failed.
.SH EXAMPLE
Export log messages from database to JSON file:
.IP
.EX
$ dmexport \-d log.db \-t log \-f json \-N dummy\-node \(rs
  \-B 2020\-01\-01 \-E 2025\-01\-01 \-o /tmp/log.json
.EE
.PP
Export observations from database to CSV file:
.IP
.EX
$ dmexport \-d observ.db \-t observ \-f csv \-N dummy\-node \(rs
  \-S dummy\-sensor \-T dummy\-target \-B 2020\-01\-01 \-E 2025\-01\-01 \(rs
  \-o /tmp/observ.csv
.EE
.SH SEE ALSO
\f[I]dmimport(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmexport.f90
! dmexport.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmexport
    !! Exports nodes, observations, sensors, targets, and logs from database to
    !! file, in ASCII block, CSV, JSON, JSON Lines, NML, or TSV format.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmexport'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    type :: app_type
        !! Command-line arguments.
        character(FILE_PATH_LEN)     :: database  = ' '         !! Path to database.
        character(FILE_PATH_LEN)     :: output    = ' '         !! Output file path, empty or '-' for stdout.
        character(NODE_ID_LEN)       :: node_id   = ' '         !! Node id.
        character(SENSOR_ID_LEN)     :: sensor_id = ' '         !! Sensor id.
        character(TARGET_ID_LEN)     :: target_id = ' '         !! Target id.
        character(TIME_LEN)          :: from      = ' '         !! Time range start.
        character(TIME_LEN)          :: to        = ' '         !! Time range end.
        character(RESPONSE_NAME_LEN) :: response  = ' '         !! Response name.
        integer                      :: format    = FORMAT_NONE !! Output format.
        integer                      :: type      = TYPE_NONE   !! Entity type.
        logical                      :: header    = .false.     !! CSV header.
        character                    :: separator = ','         !! CSV separator character.
    end type app_type

    integer        :: rc  ! Return code.
    type(app_type) :: app ! App settings.

    ! Initialise DMPACK.
    call dm_init()

    ! Read command-line arguments.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Create selected database type.
    rc = export(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)
contains
    integer function export(app) result(rc)
        type(app_type), intent(inout) :: app

        integer :: stat, unit
        logical :: empty, first, is_file

        type(db_type)      :: db
        type(db_stmt_type) :: dbs
        type(serial_class) :: serial

        type(beat_type)   :: beat
        type(dp_type)     :: dp
        type(log_type)    :: log
        type(node_type)   :: node
        type(observ_type) :: observ
        type(sensor_type) :: sensor
        type(target_type) :: target

        is_file = (dm_string_has(app%output) .and. app%output /= '-')

        ! Open file.
        unit = stdout

        if (is_file) then
            rc = E_IO
            open (action='write', file=trim(app%output), iostat=stat, newunit=unit, status='replace')
            if (stat /= 0) return
        end if

        ! Open database.
        rc = dm_db_open(db, app%database, read_only=.true., validate=.true.)

        if (rc == E_INVALID) then
            call dm_error_out(rc, 'invalid database')
            return
        end if

        if (dm_is_error(rc)) then
            call dm_error_out(rc, 'failed to open database')
            return
        end if

        ! Export rows.
        first = .true.

        do
            empty = (rc == E_DB_NO_ROWS)

            select case (app%type)
                case (TYPE_NODE)
                    rc = dm_db_select(db, dbs, node, validate=first)
                    if (first) call serial%create(node, app%format, unit=unit, empty=empty, header=app%header, separator=app%separator)
                    if (rc /= E_NONE) exit
                    call serial%next(node)

                case (TYPE_SENSOR)
                    rc = dm_db_select(db, dbs, sensor, validate=first)
                    if (first) call serial%create(sensor, app%format, unit=unit, empty=empty, header=app%header, separator=app%separator)
                    if (rc /= E_NONE) exit
                    call serial%next(sensor)

                case (TYPE_TARGET)
                    rc = dm_db_select(db, dbs, target, validate=first)
                    if (first) call serial%create(target, app%format, unit=unit, empty=empty, header=app%header, separator=app%separator)
                    if (rc /= E_NONE) exit
                    call serial%next(target)

                case (TYPE_OBSERV)
                    rc = dm_db_select(db, dbs, observ, node_id=app%node_id, sensor_id=app%sensor_id, target_id=app%target_id, from=app%from, to=app%to, validate=first)
                    if (first) call serial%create(observ, app%format, unit=unit, empty=empty, header=app%header, separator=app%separator)
                    if (rc /= E_NONE) exit
                    call serial%next(observ)

                case (TYPE_LOG)
                    rc = dm_db_select(db, dbs, log, node_id=app%node_id, sensor_id=app%sensor_id, target_id=app%target_id, from=app%from, to=app%to, validate=first)
                    if (first) call serial%create(log, app%format, unit=unit, empty=empty, header=app%header, separator=app%separator)
                    if (rc /= E_NONE) exit
                    call serial%next(log)

                case (TYPE_BEAT)
                    rc = dm_db_select(db, dbs, beat, validate=first)
                    if (first) call serial%create(beat, app%format, unit=unit, empty=empty, header=app%header, separator=app%separator)
                    if (rc /= E_NONE) exit
                    call serial%next(beat)

                case (TYPE_DP)
                    rc = dm_db_select(db, dbs, dp, node_id=app%node_id, sensor_id=app%sensor_id, target_id=app%target_id, response_name=app%response, from=app%from, to=app%to, validate=first)
                    if (first .and. app%format /= FORMAT_BLOCK) call serial%create(dp, app%format, unit=unit, empty=empty, header=app%header, separator=app%separator)
                    if (rc /= E_NONE) exit

                    if (app%format == FORMAT_BLOCK) then
                        stat = dm_block_write(dp, unit)
                    else
                        call serial%next(dp)
                    end if
            end select

            first = .false.

            if (rc == E_DB_DONE) exit

            if (dm_is_error(rc)) then
                call dm_error_out(rc, 'failed to select from database')
                exit
            end if
        end do

        call serial%finalize()
        call dm_db_finalize(dbs)
        call dm_db_close(db)

        if (is_file) close (unit)

        if (dm_is_error(rc)) then
            call dm_error_out(rc, 'failed to write data')
            return
        end if
    end function export

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments.
        type(app_type), intent(out) :: app

        character(6)           :: format_name, type_name
        type(arg_parser_class) :: parser

        call parser%add('database',  short='d', type=ARG_TYPE_DATABASE, required=.true., exist=.true.)          ! -d, --database <path>
        call parser%add('output',    short='o', type=ARG_TYPE_FILE)                                             ! -o, --output <path>
        call parser%add('node',      short='N', type=ARG_TYPE_ID)                                               ! -N, --node <id>
        call parser%add('sensor',    short='S', type=ARG_TYPE_ID)                                               ! -S, --sensor <id>
        call parser%add('target',    short='T', type=ARG_TYPE_ID)                                               ! -T, --target <id>
        call parser%add('from',      short='B', type=ARG_TYPE_TIME)                                             ! -F, --from <timestamp>
        call parser%add('to',        short='E', type=ARG_TYPE_TIME)                                             ! -T, --to <timestamp>
        call parser%add('response',  short='R', type=ARG_TYPE_ID,     max_len=RESPONSE_NAME_LEN)                ! -R, --response <name>
        call parser%add('format',    short='f', type=ARG_TYPE_STRING, max_len=FORMAT_NAME_LEN, required=.true.) ! -f, --format <string>
        call parser%add('type',      short='t', type=ARG_TYPE_STRING, max_len=TYPE_NAME_LEN,   required=.true.) ! -t, --type <string>
        call parser%add('header',    short='H', type=ARG_TYPE_LOGICAL)                                          ! -H, --header
        call parser%add('separator', short='s', type=ARG_TYPE_CHAR)                                             ! -a, --separator <char>

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('database',  app%database)
        call parser%get('output',    app%output)
        call parser%get('node',      app%node_id)
        call parser%get('sensor',    app%sensor_id)
        call parser%get('target',    app%target_id)
        call parser%get('from',      app%from)
        call parser%get('to',        app%to)
        call parser%get('response',  app%response)
        call parser%get('format',    format_name)
        call parser%get('type',      type_name)
        call parser%get('header',    app%header)
        call parser%get('separator', app%separator)

        app%format = dm_format_from_name(format_name)
        app%type   = dm_type_from_name(type_name)

        rc = validate(app)
    end function read_args

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        ! Serialisation format.
        select case (app%format)
            case (FORMAT_BLOCK, FORMAT_CSV, FORMAT_JSON, FORMAT_JSONL, FORMAT_NML, FORMAT_TSV)
                continue
            case default
                call dm_error_out(rc, 'invalid format')
                return
        end select

        ! Data type.
        select case (app%type)
            case (TYPE_NODE, TYPE_SENSOR, TYPE_TARGET, TYPE_OBSERV, TYPE_LOG, TYPE_BEAT, TYPE_DP)
                continue
            case default
                call dm_error_out(rc, 'invalid type')
                return
        end select

        ! Log, observation, and data point.
        if (app%type == TYPE_LOG .or. app%type == TYPE_OBSERV .or. app%type == TYPE_DP) then
            if (.not. dm_time_is_valid(app%from)) then
                call dm_error_out(rc, 'invalid or missing argument --from')
                return
            end if

            if (.not. dm_time_is_valid(app%to)) then
                call dm_error_out(rc, 'invalid or missing argument --to')
                return
            end if
        end if

        ! Observation and data point.
        if (app%type == TYPE_OBSERV .or. app%type == TYPE_DP) then
            if (.not. dm_id_is_valid(app%node_id)) then
                 call dm_error_out(rc, 'invalid or missing argument --node')
                 return
            end if

            if (.not. dm_id_is_valid(app%sensor_id)) then
                call dm_error_out(rc, 'invalid or missing argument --sensor')
                return
            end if

            if (.not. dm_id_is_valid(app%target_id)) then
                call dm_error_out(rc, 'invalid or missing argument --target')
                return
            end if
        end if

        ! Data point.
        if (app%type == TYPE_DP .and. .not. dm_id_is_valid(app%response)) then
            call dm_error_out(rc, 'invalid or missing argument --response')
            return
        end if

        ! Block format.
        if (app%type /= TYPE_DP .and. app%format == FORMAT_BLOCK) then
            call dm_error_out(rc, 'block format not supported')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_db_version(.true.)
    end subroutine version_callback
end program dmexport
